helix-core/src/main/java/org/apache/helix/model/ControllerHistory.java

```diff
@@ -136,6 +134,40 @@ public ZNRecord updateHistory(String clusterName, String instanceName, String ve
     return historyList;
   }
 
+  /**
+   * Gets the management mode history.
+   *
+   * @return List of history json strings.
+   */
+  public List<String> getManagementModeHistory() {
+    List<String> history =
+        _record.getListField(ManagementModeConfigKey.MANAGEMENT_MODE_HISTORY.name());
+    return history == null ? Collections.emptyList() : history;
+  }
+
+  /**
+   * Updates management mode and status history to controller history in FIFO order.
+   *
+   * @param controller controller name
+   * @param mode cluster management mode {@link ClusterManagementMode}
+   * @param fromHost the hostname that creates the management mode signal
+   * @param time time in millis
+   * @param reason reason to put the cluster in management mode
+   * @return updated history znrecord
+   */
+  public ZNRecord updateManagementModeHistory(String controller, ClusterManagementMode mode,
+      String fromHost, long time, String reason) {
+    Map<String, String> historyEntry = new HashMap<>();
+    historyEntry.put(ConfigProperty.CONTROLLER.name(), controller);
+    historyEntry.put(ConfigProperty.TIME.name(), Instant.ofEpochMilli(time).toString());
+    historyEntry.put(ManagementModeConfigKey.MODE.name(), mode.getMode().name());
+    historyEntry.put(ManagementModeConfigKey.STATUS.name(), mode.getStatus().name());
```

Comment:
So the status would always be completed, and it is not delivering any useful information, right?

Reply:
We (you and me) discussed it offline before and agreed on recording COMPLETED is good enough. For sure we could record STARTED as well if we think it is very useful.
The status field makes it clear to users the timestamp is for completed. Otherwise, if users have no context, they would be confused: what does it mean. So I prefer to make it clear. And it's extensible for use to add other status, eg. STARTED.

Reply:
Thanks for the details. I don't see any real problem here. Although I still don't think the additional status string is useful, please feel free to keep it if you believe it can provide value.